mqttclients/dotnet/MQTTnet.Client.Extensions/X509ChainValidator.cs

```diff
@@ -1,33 +1,70 @@
-            chain.ChainPolicy.UrlRetrievalTimeout = new TimeSpan(0, 0, 0);
-            if (certChain != null)
+            X509Certificate2Collection caCerts = new();
+            if (!string.IsNullOrEmpty(caCertFile))
             {
-                chain.ChainPolicy.CustomTrustStore.AddRange(certChain);
-                chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
+                caCerts.ImportFromPemFile(caCertFile);
             }
-            X509Certificate2 x5092 = new X509Certificate2(cert);
-            bool res = chain.Build(x5092);
-            if (res == false)
+            return ValidateChain(certValArgs, caCerts);
+        }
+
+        internal static bool ValidateChain(MqttClientCertificateValidationEventArgs certValArgs, X509Certificate2Collection caChain)
+        {
+            if (certValArgs.SslPolicyErrors == SslPolicyErrors.None)
             {
-                Trace.TraceError($"Error validating TLS chain for cert: '{cert.Subject}' issued by '{cert.Issuer}'");
-                chain.ChainStatus.ToList().ForEach(s => Trace.TraceError(s.StatusInformation));
+                return true;
             }
-            return res;
-        }
 
+            Trace.TraceError($"SslPolicyErrors: '{certValArgs.SslPolicyErrors}'");
+            Trace.TraceWarning($"Loaded {caChain.Count} certs ");
+            caChain.ToList().ForEach(c => Trace.TraceWarning(c.Subject));
+
+            var sslErrors = Enum.GetValues<SslPolicyErrors>().Where(e => certValArgs.SslPolicyErrors.HasFlag(e)).ToList();
+            bool valid = false;
 
+            foreach (var sslError in sslErrors)
+            {
+                if (sslError == SslPolicyErrors.RemoteCertificateChainErrors)
+                {
+                    bool chainValidated = false;
+                    X509Chain chain = new();
+                    chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
+                    chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+                    chain.ChainPolicy.VerificationTime = DateTime.UtcNow;
+                    chain.ChainPolicy.UrlRetrievalTimeout = new TimeSpan(0, 0, 10);
+                    chain.ChainPolicy.CustomTrustStore.AddRange(caChain);
+                    chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
+
+                    X509Certificate cert = certValArgs.Certificate;
+                    X509Certificate2 x5092 = new(cert);
+                    chainValidated = chain.Build(x5092);
+                    if (chainValidated == false)
+                    {
+                        Trace.TraceError($"Error validating TLS chain for cert: '{cert.Subject}' issued by '{cert.Issuer}'");
+                        chain.ChainStatus.ToList().ForEach(s => Trace.TraceError(s.StatusInformation));
+                    }
+                    valid = chainValidated;
+                }
+                if (sslError == SslPolicyErrors.RemoteCertificateNotAvailable)
```

Comment:
nit: consider combining these if statements into one, if it's easy to read that way

Reply:
how? we need to invoke `Build` in the first if